### PR TITLE
wwan: adding cdc_ether modem support

### PR DESCRIPTION
**Pull request by fid0did0**
Some modem use the USB cdc_ether interface to transfer data and the usb_serial to control the modem.
This patch provides support for these modem.

Signed-off-by: Giuseppe Lippolis <giu.lippolis@gmail.com>

### CODE REVIEW

**Comment by fid0did0** [comment]
Using the current patch I'm able to establish a data connection on the cdc_ether usb interface:

```
root@lede:~# logread|grep wan3g
Tue Nov 29 22:30:58 2016 daemon.notice netifd: Interface 'wan3g' is setting up now
Tue Nov 29 22:30:58 2016 daemon.notice netifd: wan3g (1113): wwan[1113] Using proto:wwan device:/dev/ttyUSB0 iface:usb0 desc:
Tue Nov 29 22:31:02 2016 daemon.notice netifd: wan3g (1113): sending -> AT+CFUN=1
Tue Nov 29 22:31:05 2016 daemon.notice netifd: wan3g (1113): Trying to set PIN
Tue Nov 29 22:31:06 2016 daemon.notice netifd: wan3g (1113): PIN set successfully
Tue Nov 29 22:31:06 2016 daemon.notice netifd: wan3g (1113): wwan[1113] DEBUG 2+
Tue Nov 29 22:31:07 2016 daemon.notice netifd: wan3g (1113): sending -> AT+CREG?
Tue Nov 29 22:31:07 2016 daemon.notice netifd: wan3g (1113): sending -> AT+CGDCONT=1,"IP","web.vodafone.de",0,0
Tue Nov 29 22:31:09 2016 daemon.notice netifd: wan3g (1113): sending -> AT+CGACT=1,1
Tue Nov 29 22:31:12 2016 daemon.notice netifd: wan3g (1113): Connected, starting DHCP on usb0
Tue Nov 29 22:31:12 2016 daemon.notice netifd: Interface 'wan3g' is now up
Tue Nov 29 22:31:13 2016 daemon.notice netifd: wan3g (1113): sending -> AT+CGDATA="M-MBIM",1,1
Tue Nov 29 22:31:14 2016 daemon.notice netifd: Interface 'wan3g_4' is enabled
Tue Nov 29 22:31:14 2016 daemon.notice netifd: Interface 'wan3g_4' has link connectivity
Tue Nov 29 22:31:14 2016 daemon.notice netifd: Interface 'wan3g_4' is setting up now
Tue Nov 29 22:31:14 2016 daemon.notice netifd: wan3g (1113): Command failed: Unknown error
Tue Nov 29 22:31:14 2016 daemon.notice netifd: wan3g_4 (1209): udhcpc: started, v1.25.1
Tue Nov 29 22:31:14 2016 daemon.notice netifd: wan3g_4 (1209): udhcpc: sending discover
Tue Nov 29 22:31:17 2016 daemon.notice netifd: wan3g_4 (1209): udhcpc: sending discover
Tue Nov 29 22:31:17 2016 daemon.notice netifd: wan3g_4 (1209): udhcpc: sending select for 2.205.215.60
Tue Nov 29 22:31:17 2016 daemon.notice netifd: wan3g_4 (1209): udhcpc: lease of 2.205.215.60 obtained, lease time 268435455
Tue Nov 29 22:31:17 2016 daemon.notice netifd: Interface 'wan3g_4' is now up

root@lede:~# ifconfig
usb0      Link encap:Ethernet  HWaddr 02:00:FF:AA:AA:AA
          inet addr:2.205.215.60  Bcast:2.205.215.63  Mask:255.255.255.248
          inet6 addr: fe80::ffff:feaa:aaaa/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:42 errors:0 dropped:0 overruns:0 frame:0
          TX packets:43 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:4050 (3.9 KiB)  TX bytes:4455 (4.3 KiB)

```
Here the usb interface description:
```
T:  Bus=01 Lev=01 Prnt=01 Port=00 Cnt=01 Dev#=  2 Spd=480  MxCh= 0
D:  Ver= 2.00 Cls=02(comm.) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=2001 ProdID=7d04 Rev= 3.00
S:  Manufacturer=D-Link,Inc
S:  Product=D-Link DWM-158
C:* #Ifs= 7 Cfg#= 1 Atr=a0 MxPwr=500mA
A:  FirstIf#= 0 IfCount= 2 Cls=02(comm.) Sub=06 Prot=00
I:* If#= 0 Alt= 0 #EPs= 1 Cls=02(comm.) Sub=06 Prot=00 Driver=cdc_ether
E:  Ad=88(I) Atr=03(Int.) MxPS=  64 Ivl=125us
I:  If#= 1 Alt= 0 #EPs= 0 Cls=0a(data ) Sub=00 Prot=00 Driver=cdc_ether
I:* If#= 1 Alt= 1 #EPs= 2 Cls=0a(data ) Sub=00 Prot=00 Driver=cdc_ether
E:  Ad=81(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=01(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 2 Alt= 0 #EPs= 3 Cls=ff(vend.) Sub=02 Prot=01 Driver=option
E:  Ad=87(I) Atr=03(Int.) MxPS=  64 Ivl=125us
E:  Ad=82(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=02(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 3 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=00 Prot=00 Driver=option
E:  Ad=83(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=03(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 4 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=00 Prot=00 Driver=option
E:  Ad=84(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=04(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 5 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=00 Prot=00 Driver=option
E:  Ad=85(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=05(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 6 Alt= 0 #EPs= 2 Cls=08(stor.) Sub=06 Prot=50 Driver=(none)
E:  Ad=86(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=06(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
```

File: package/network/utils/comgt/files/ncm.sh [hunk 1]
```diff
@@ -84,6 +84,7 @@ proto_ncm_setup() {
 	[ -n "$delay" ] && sleep "$delay"
 
 	manufacturer=`gcom -d "$device" -s /etc/gcom/getcardinfo.gcom | awk '/Manufacturer/ { print tolower($2) }'`
+	manufacturer=`gcom -d "$device" -s /etc/gcom/getcardinfo.gcom | head -2 | tail -1 | tr -d "\r" | tr [A-Z] [a-z]`
```
Comment (blogic) [inline package/network/utils/comgt/files/ncm.sh]: explain this line please and why is the line above not getting removed ?

File: package/network/utils/comgt/files/ncm.sh [hunk 2]
```diff
@@ -126,18 +137,27 @@ proto_ncm_setup() {
...
-		echo "Failed to connect"
-		proto_notify_error "$interface" CONNECT_FAILED
-		return 1
-	}
+#	json_get_vars connect
+#	eval COMMAND="$connect" gcom -d "$device" -s /etc/gcom/runcommand.gcom || {
+#		echo "Failed to connect"
+#		proto_notify_error "$interface" CONNECT_FAILED
+#		return 1
+#	}
```
Comment (blogic) [inline package/network/utils/comgt/files/ncm.sh]: dont leave dead code in the file

File: package/network/utils/comgt/files/ncm.sh [hunk 3]
```diff
@@ -126,18 +137,27 @@ proto_ncm_setup() {
...
 
 	echo "Connected, starting DHCP on $ifname"
 	
 	proto_init_update "$ifname" 1
 	proto_send_update "$interface"
 
+	son_load "$(cat /etc/gcom/ncm.json)"
+	son_select "$manufacturer"
+	son_get_vars connect
+	val COMMAND="$connect" gcom -d "$device" -s /etc/gcom/runcommand.gcom || {
```
Comment (blogic) [inline package/network/utils/comgt/files/ncm.sh]: untested code. i doubt you runtime tested this otherwise you would have noticed the missing "j" in "son"

File: package/network/utils/comgt/files/ncm.sh [hunk 4]
```diff
@@ -164,9 +184,10 @@ proto_ncm_teardown() {
 	local device
 	json_get_vars device
 
-	echo "Stopping network"
+	echo "Stopping network device:$device"
 
 	manufacturer=`gcom -d "$device" -s /etc/gcom/getcardinfo.gcom | awk '/Manufacturer/ { print tolower($2) }'`
+	manufacturer=`gcom -d "$device" -s /etc/gcom/getcardinfo.gcom | sed '2q;d' | tr -d "\n\r" | tr [A-Z] [a-z]`
```
Comment (blogic) [inline package/network/utils/comgt/files/ncm.sh]: same as before, done duplicate and leave the old line in place. please explain the change

File: package/network/utils/comgt/files/ncm.json [hunk 5]
```diff
@@ -63,5 +63,21 @@
 		},
 		"connect": "AT*ENAP=1,1",
 		"disconnect": "AT*ENAP=0"
+	},
+	"mtk2": {
+                "initialize": [
+			"AT+CFUN=1"
+		],
+                "configure": [
+			"AT+CREG?",
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: Is this command really necessary? It shouldn't really do anything due to the "?" at the end.

File: package/network/utils/comgt/files/ncm.json [hunk 6]
```diff
@@ -63,5 +63,21 @@
 		},
 		"connect": "AT*ENAP=1,1",
 		"disconnect": "AT*ENAP=0"
+	},
+	"mtk2": {
+                "initialize": [
+			"AT+CFUN=1"
+		],
+                "configure": [
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: Why can't these be under "initialize"?

File: package/network/utils/comgt/files/ncm.json [hunk 7]
```diff
@@ -63,5 +63,21 @@
...
+			"AT+CREG?",
+			"AT+CGDCONT=1,\\\"${pdptype}\\\",\\\"${apn}\\\",0,0",
+                        "AT+CGACT=1,1"
+                ],
+                "modes": {
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: Leave these out if you don't know the correct commands.

File: package/network/utils/comgt/files/ncm.sh (hunk 1 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: Tell us what the output of "ATI" is. We probably need to modify the manufacturer detection for other models as well.

File: package/network/utils/comgt/files/ncm.sh [hunk 8]
```diff
@@ -115,6 +116,16 @@ proto_ncm_setup() {
...
+
+        json_get_values configure configure
+        for i in $configure; do
+                eval COMMAND="$i" gcom -d "$device" -s /etc/gcom/runcommand.gcom || {
+                        echo "Failed to configure modem"
+                        proto_notify_error "$interface" INITIALIZE_FAILED
+                        return 1
+                }
+        done
+
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: Again, why do you need this separate "configure" block?

File: package/network/utils/comgt/files/ncm.sh (hunk 3 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: Yeah, and I really doubt that you need to actually move this block here.

File: package/network/utils/comgt/files/ncm.sh [hunk 9]
```diff
@@ -164,9 +184,10 @@ proto_ncm_teardown() {
 	local device
 	json_get_vars device
 
-	echo "Stopping network"
+	echo "Stopping network device:$device"
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: This change is unrelated.

File: package/network/utils/comgt/files/ncm.json [hunk 10]
```diff
@@ -63,5 +63,21 @@
...
 		"disconnect": "AT*ENAP=0"
+	},
+	"mtk2": {
+                "initialize": [
+			"AT+CFUN=1"
+		],
+                "configure": [
+			"AT+CREG?",
+			"AT+CGDCONT=1,\\\"${pdptype}\\\",\\\"${apn}\\\",0,0",
+                        "AT+CGACT=1,1"
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: Why can't these be under "initialize"?

File: package/network/utils/comgt/files/ncm.json [hunk 11]
```diff
@@ -63,5 +63,21 @@
...
+                ],
+                "modes": {
+                        "umts": "AT",
+                        "gsm": "AT"
+                },
```
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: Leave these out if you don't know the correct commands.

File: package/network/utils/comgt/files/ncm.sh (hunk 3 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: And I really doubt that you need to actually move the "connect" block here.

File: package/network/utils/comgt/files/ncm.sh (hunk 1 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: I just sent a related patch to the mailing list, where among other changes the AT+CGMI command is used for identifying the modem. I think it should be straightforward to use that.

File: package/network/utils/comgt/files/ncm.json (hunk 10 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.json]: I see, that makes sense. Other modems may have similar restrictions, I just always disable the PIN query. Adding a separate configure block is a good idea.

File: package/network/utils/comgt/files/ncm.sh (hunk 2 above)
Comment (malaakso) [inline package/network/utils/comgt/files/ncm.sh]: I don't yet see what is the need to move this block? Your change shouldn't have any effect.

Comment (fid0did0) [inline package/network/utils/comgt/files/ncm.sh]: You are right, this change require an explanation.
The connection cmd need to be sent after the interface is enabled.
This is the reason why I moved the section after the:
```
proto_init_update "$ifname" 1
proto_send_update "$interface"
```
I will show the failure log in the next day.

**Comment by fid0did0** [comment]
After the discussion with Matti, I would keep this patch only to help myself to fix the problem I currently have on my specific modem model (in the proto_ncm_teardown function).
At the end I will close this patch and, in case, I will open a new one, merging the Matti's update.

**Comment by blogic** [comment]
@fid0did0 can i close it in that case ?

**Comment by fid0did0** [comment]
Ok john I currently close this PR and I will open again after the integration of the work from Matti.
